File: src/lib/encrypt.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto"
import consola from "consola"

const ALGORITHM = "aes-256-gcm"
const ENC_PREFIX = "enc:"

let encryptionKey: Buffer | null = null

/**
 * 初始化加密密钥，从 ENCRYPTION_KEY 环境变量读取 64 位 hex 字符串（32 字节）
 * 若未设置则发出警告并以明文模式运行；若格式错误则退出进程。
 */
export function initEncryption(): void {
  const keyHex = process.env.ENCRYPTION_KEY
  if (!keyHex) {
    consola.error("ENCRYPTION_KEY 未设置 — 拒绝启动以防止敏感数据明文存储。")
    consola.error("请在 .env 中设置 ENCRYPTION_KEY=<64 位 hex 字符串>（可用 openssl rand -hex 32 生成）。")
    process.exit(1)
  }
  if (!/^[0-9a-fA-F]{64}$/.test(keyHex)) {
    consola.error("ENCRYPTION_KEY 必须是 64 个十六进制字符（32 字节）")
    process.exit(1)
  }
  encryptionKey = Buffer.from(keyHex, "hex")
  consola.info("静态加密已启用（AES-256-GCM）")
}

/**
 * 加密字符串，返回格式：enc:<iv_hex>:<authTag_hex>:<ciphertext_hex>
 * 若未设置 ENCRYPTION_KEY 或值已加密则原样返回。
 */
export function encrypt(plaintext: string): string {
  if (!encryptionKey || plaintext.startsWith(ENC_PREFIX)) return plaintext
  const iv = randomBytes(12) // 96-bit IV for GCM
  const cipher = createCipheriv(ALGORITHM, encryptionKey, iv)
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()])
  const authTag = cipher.getAuthTag()
  return `${ENC_PREFIX}${iv.toString("hex")}:${authTag.toString("hex")}:${ciphertext.toString("hex")}`
}

/**
 * 解密字符串，支持：
 * - 已加密值（enc:... 前缀）→ 解密
 * - 明文（无前缀）→ 原样返回（向后兼容）
 */
export function decrypt(value: string): string {
  if (!value.startsWith(ENC_PREFIX)) return value
  if (!encryptionKey) {
    consola.error("发现加密数据但 ENCRYPTION_KEY 未设置，无法解密，请在 .env 中配置密钥。")
    return value
  }
  const parts = value.slice(ENC_PREFIX.length).split(":")
  if (parts.length !== 3) {
    consola.warn("加密值格式不合法，返回原始值")
    return value
  }
  const [ivHex, tagHex, ctHex] = parts
  try {
    const decipher = createDecipheriv(ALGORITHM, encryptionKey, Buffer.from(ivHex, "hex"))
    decipher.setAuthTag(Buffer.from(tagHex, "hex"))
    return Buffer.concat([
      decipher.update(Buffer.from(ctHex, "hex")),
      decipher.final(),
    ]).toString("utf8")
  } catch {
    consola.error("解密失败 — 密钥错误或数据损坏")
    return value
  }
}

File: backend/internal/crypto/encrypt.go
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"log"
	"os"
	"regexp"
	"strings"
)

const encPrefix = "enc:"

var encryptionKey []byte

// InitEncryption initializes the AES-256-GCM encryption key from ENCRYPTION_KEY env var.
// Exits the process if the key is not set or has an invalid format.
func InitEncryption() {
	keyHex := os.Getenv("ENCRYPTION_KEY")
	if keyHex == "" {
		log.Println("ERROR: ENCRYPTION_KEY 未设置 — 拒绝启动以防止敏感数据明文存储。")
		log.Println("ERROR: 请设置 ENCRYPTION_KEY=<64 位 hex 字符串>（可用 openssl rand -hex 32 生成）。")
		os.Exit(1)
	}
	matched, _ := regexp.MatchString(`^[0-9a-fA-F]{64}$`, keyHex)
	if !matched {
		log.Println("ERROR: ENCRYPTION_KEY 必须是 64 个十六进制字符（32 字节）")
		os.Exit(1)
	}
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		log.Printf("ERROR: ENCRYPTION_KEY 解码失败: %v", err)
		os.Exit(1)
	}
	encryptionKey = key
	log.Println("INFO: 静态加密已启用（AES-256-GCM）")
}

// Encrypt encrypts a plaintext string using AES-256-GCM.
// Returns format: enc:<iv_hex>:<authTag_hex>:<ciphertext_hex>
// If already encrypted (starts with enc:), returns as-is.
func Encrypt(plaintext string) string {
	if encryptionKey == nil || strings.HasPrefix(plaintext, encPrefix) {
		return plaintext
	}

	block, err := aes.NewCipher(encryptionKey)
	if err != nil {
		log.Printf("ERROR: 创建 cipher 失败: %v", err)
		return plaintext
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		log.Printf("ERROR: 创建 GCM 失败: %v", err)
		return plaintext
	}

	iv := make([]byte, 12) // 96-bit IV for GCM
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		log.Printf("ERROR: 生成 IV 失败: %v", err)
		return plaintext
	}

	// gcm.Seal appends ciphertext+tag (tag is last 16 bytes)
	encrypted := gcm.Seal(nil, iv, []byte(plaintext), nil)
	ciphertext := encrypted[:len(encrypted)-16]
	tag := encrypted[len(encrypted)-16:]

	return fmt.Sprintf("%s%x:%x:%x", encPrefix, iv, tag, ciphertext)
}

// Decrypt decrypts a value encrypted with Encrypt.
// Supports backward-compatible plaintext values (no enc: prefix).
func Decrypt(value string) string {
	if !strings.HasPrefix(value, encPrefix) {
		return value // plaintext, backward compatible
	}
	if encryptionKey == nil {
		log.Println("ERROR: 发现加密数据但 ENCRYPTION_KEY 未设置，无法解密")
		return value
	}

	parts := strings.SplitN(value[len(encPrefix):], ":", 3)
	if len(parts) != 3 {
		log.Println("WARN: 加密值格式不合法，返回原始值")
		return value
	}

	ivBytes, err1 := hex.DecodeString(parts[0])
	tagBytes, err2 := hex.DecodeString(parts[1])
	ctBytes, err3 := hex.DecodeString(parts[2])
	if err1 != nil || err2 != nil || err3 != nil {
		log.Println("WARN: 加密值 hex 解码失败")
		return value
	}

	block, err := aes.NewCipher(encryptionKey)
	if err != nil {
		log.Printf("ERROR: 创建 cipher 失败: %v", err)
		return value
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		log.Printf("ERROR: 创建 GCM 失败: %v", err)
		return value
	}

	// TS uses createDecipheriv + setAuthTag(tag) + update(ct)
	// Go equivalent: pass ct+tag combined to gcm.Open
	combined := append(ctBytes, tagBytes...) //nolint:gocritic
	plaintext, err := gcm.Open(nil, ivBytes, combined, nil)
	if err != nil {
		log.Printf("ERROR: 解密失败 — 密钥错误或数据损坏: %v", err)
		return value
	}

	return string(plaintext)
}
